### PR TITLE
 feature/create test with parallel

### PR DESCRIPTION
**Pull request by at-hanhnguyen2**

### CODE REVIEW

File: src/test/java/vn/asiantech/base/CucumberRunnerBase.java [hunk 1]
```diff
@@ -13,30 +13,16 @@
...
-        DriverBase.closeDriverObjects();
-    }
-
     @BeforeClass(alwaysRun = true)
-    @Override
-    public void setUpClass() throws Exception {
-        super.setUpClass();
-        System.out.println("setUpClass");
+    @Parameters("browserType")
+    public void setUpClass(String browserType) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/CucumberRunnerBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/CucumberRunnerBase.java#L18"></span>Parameter browserType should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 2]
```diff
@@ -20,19 +20,17 @@
...
-    private static ThreadLocal<DriverFactory> driverFactoryThread;
+    private static ThreadLocal<DriverFactory> driverFactoryThread = new ThreadLocal<>();
 
-    static void instantiateDriverObject() {
-        driverFactoryThread = ThreadLocal.withInitial(() -> {
-            DriverFactory driverFactory = new DriverFactory();
-            webDriverThreadPool.add(driverFactory);
-            return driverFactory;
-        });
+    public static synchronized void instantiateDriverObject(String browserType) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/DriverBase.java#L25"></span>Parameter browserType should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 3]
```diff
@@ -10,7 +10,7 @@
 
 import static org.openqa.selenium.Proxy.ProxyType.MANUAL;
 import static org.openqa.selenium.remote.CapabilityType.PROXY;
-import static vn.asiantech.base.DriverType.CHROME;
+import static vn.asiantech.base.DriverType.*;
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/DriverFactory.java#L13"></span>Using the ‘.<em>’ form of import should be avoided - vn.asiantech.base.DriverType.</em>.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 4]
```diff
@@ -22,10 +22,9 @@
     private final Integer proxyPort = Integer.getInteger("proxyPort");
     private final String proxyDetails = String.format("%s:%d", proxyHostname, proxyPort);
     private RemoteWebDriver driver;
-    private DriverType selectedDriverType;
 
-    DriverFactory() {
-        DriverType driverType = CHROME;
+    public DriverFactory(String browserType) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/DriverFactory.java#L26"></span>Parameter browserType should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 5]
```diff
@@ -34,17 +33,42 @@
...
     }
 
-    public RemoteWebDriver getDriver() {
-        if (null == driver) {
-            try {
-                instantiateWebDriver(selectedDriverType);
-            } catch (MalformedURLException e) {
-                e.printStackTrace();
-            }
+    private DriverType getDriverType(String browserType) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/DriverFactory.java#L44"></span>Parameter browserType should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 6]
```diff
@@ -34,17 +33,42 @@
...
+            case Constant.BROWSER_OPERA:
+                driverType = OPERA;
+                break;
+            default:
+                driverType = CHROME;
         }
+        return driverType;
+    }
+
+    public RemoteWebDriver getDriver() {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/dd2c93731070e79ff5d25593155b1d3d9844f01e/src/test/java/vn/asiantech/base/DriverFactory.java#L71"></span>Method ‘getDriver’ is not designed for extension - needs to be abstract, final or empty.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/Constant.java [hunk 7]
```diff
@@ -14,6 +16,24 @@
...
 
+    static final String BROWSER_CHROME = "chrome";
+    static final String BROWSER_FIREFOX = "firefox";
+    static final String BROWSER_SAFARI = "safari";
+    static final String BROWSER_IE = "ie";
+    static final String BROWSER_EDGE = "edge";
+    static final String BROWSER_OPERA = "opera";
+
+    //init Account
+    static final Account[] accountLogin = new Account[]{
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/Constant.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0be637c2e534f456e0ab2ca0e0f067d3dafbb58f/src/test/java/vn/asiantech/base/Constant.java#L27"></span>Name ‘accountLogin’ must match pattern ‘^[A-Z][A-Z0-9]<em>(_[A-Z0-9]+)</em>$’.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 8]
```diff
@@ -20,23 +21,25 @@
...
-        if (driverFactoryThread == null) {
-            instantiateDriverObject();
+    public static synchronized RemoteWebDriver getDriver() {
+        if (driverFactoryThread.get() == null) {
+            instantiateDriverObject(Constant.BROWSER_CHROME);
         }
         return driverFactoryThread.get().getDriver();
     }
 
+    protected Account getAccount() {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0be637c2e534f456e0ab2ca0e0f067d3dafbb58f/src/test/java/vn/asiantech/base/DriverBase.java#L39"></span>Method ‘getAccount’ is not designed for extension - needs to be abstract, final or empty.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 9]
```diff
@@ -4,13 +4,16 @@
...
 
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
 
 import static org.openqa.selenium.Proxy.ProxyType.MANUAL;
 import static org.openqa.selenium.remote.CapabilityType.PROXY;
-import static vn.asiantech.base.DriverType.CHROME;
+import static vn.asiantech.base.DriverType.*;
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0be637c2e534f456e0ab2ca0e0f067d3dafbb58f/src/test/java/vn/asiantech/base/DriverFactory.java#L16"></span>Using the ‘.<em>’ form of import should be avoided - vn.asiantech.base.DriverType.</em>.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 10]
```diff
@@ -22,10 +25,28 @@
...
     private final String proxyDetails = String.format("%s:%d", proxyHostname, proxyPort);
     private RemoteWebDriver driver;
-    private DriverType selectedDriverType;
 
-    DriverFactory() {
-        DriverType driverType = CHROME;
+    private static List<Integer> busyAccounts = new ArrayList<>();
+    private Account accountCanUse;
+
+    Account getAccountCanUse() {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0be637c2e534f456e0ab2ca0e0f067d3dafbb58f/src/test/java/vn/asiantech/base/DriverFactory.java#L32"></span>Method ‘getAccountCanUse’ is not designed for extension - needs to be abstract, final or empty.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 11]
```diff
@@ -34,17 +55,42 @@
...
     }
 
-    public RemoteWebDriver getDriver() {
-        if (null == driver) {
-            try {
-                instantiateWebDriver(selectedDriverType);
-            } catch (MalformedURLException e) {
-                e.printStackTrace();
-            }
+    private DriverType getDriverType(String browserType) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0be637c2e534f456e0ab2ca0e0f067d3dafbb58f/src/test/java/vn/asiantech/base/DriverFactory.java#L66"></span>Parameter browserType should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 12]
```diff
@@ -34,17 +54,42 @@
...
     }
 
-    public RemoteWebDriver getDriver() {
-        if (null == driver) {
-            try {
-                instantiateWebDriver(selectedDriverType);
-            } catch (MalformedURLException e) {
-                e.printStackTrace();
-            }
+    private DriverType getDriverType(String browserName) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/1382743b99762e8359c40f9b94b6cd70631651f5/src/test/java/vn/asiantech/base/DriverFactory.java#L65"></span>Parameter browserName should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 13]
```diff
@@ -9,47 +9,55 @@
...
-    private static ThreadLocal<DriverFactory> driverFactoryThread;
+    private static ThreadLocal<DriverFactory> driverFactoryThread = new ThreadLocal<>();
 
-    static void instantiateDriverObject() {
-        driverFactoryThread = ThreadLocal.withInitial(() -> {
-            DriverFactory driverFactory = new DriverFactory();
-            webDriverThreadPool.add(driverFactory);
-            return driverFactory;
-        });
+    static synchronized void instantiateDriverObject(XmlTest xmlTest) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/1a298420407980be34610d1437dd9e6921507ee3/src/test/java/vn/asiantech/base/DriverBase.java#L25"></span>Parameter xmlTest should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 14]
```diff
@@ -1,71 +1,113 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
 import static org.openqa.selenium.Proxy.ProxyType.MANUAL;
 import static org.openqa.selenium.remote.CapabilityType.PROXY;
-import static vn.asiantech.base.DriverType.CHROME;
+import static vn.asiantech.base.DriverType.*;
 
-public class DriverFactory {
+class DriverFactory {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/1a298420407980be34610d1437dd9e6921507ee3/src/test/java/vn/asiantech/base/DriverFactory.java#L18"></span>Missing a Javadoc comment.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 15]
```diff
@@ -1,71 +1,113 @@
...
+        for (Account account : Constant.ACCOUNT_LOGIN) {
+            if (!busyAccounts.contains(account.hashCode())) {
+                busyAccounts.add(account.hashCode());
+                accountCanUse = account;
+                break;
+            }
+        }
+    }
+
+    DriverFactory(XmlTest xmlTest) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/1a298420407980be34610d1437dd9e6921507ee3/src/test/java/vn/asiantech/base/DriverFactory.java#L45"></span>Parameter xmlTest should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverFactory.java [hunk 16]
```diff
@@ -1,71 +1,116 @@
...
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/1dcf64d9bd7ad8c7fae19d87de158510f9df1979/src/test/java/vn/asiantech/base/DriverFactory.java#L48"></span>Parameter xmlTest should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/resources/DefaultSuite.xml [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+<suite name="Suite default" verbose="1">
+
+    <test name="Chrome Test">
+        <parameter name="browserName" value="chrome"/>
+        <parameter name="server" value=""/>
+        <classes>
+            <class name="TestRunner"/>
+        </classes>
+    </test>
+</suite>
```
Comment (at-hangtran) [inline src/test/resources/DefaultSuite.xml]: Please add a break line

File: src/test/resources/GridParallelSuite.xml [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    </test>
+
+    <test name="Safari Test">
+        <parameter name="browserName" value="safari"/>
+        <parameter name="server" value="http://localhost:7779/wd/hub"/>
+        <classes>
+            <class name="TestRunner"/>
+        </classes>
+    </test>
+</suite>
```
Comment (at-hangtran) [inline src/test/resources/GridParallelSuite.xml]: Please add a break line

File: src/test/resources/ParallelSuite.xml [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    </test>
+
+    <test name="Safari Test">
+        <parameter name="browserName" value="safari"/>
+        <parameter name="server" value=""/>
+        <classes>
+            <class name="TestRunner"/>
+        </classes>
+    </test>
+</suite>
```
Comment (at-hangtran) [inline src/test/resources/ParallelSuite.xml]: Please add a break line

**Pull request merged**